### PR TITLE
Upsert aka On Conflict Do Update

### LINKED ISSUE #56: PG Support Upsert
This is to prevent duplicates. Upsert is equivalent to `INSERT ... ON CONFLICT DO UPDATE`. Pandas does not directly support this with data frames but it can be implemented roughly as follows:

```py
import pandas as pd
from sqlalchemy import create_engine, text

# Create a database engine
engine = create_engine("postgresql+psycopg2://username:password@localhost:5432/mydatabase")

# Sample data
data = {'user_id': [1, 2, 3], 'username': ['alice', 'bob', 'charlie'], 'email': ['alice@example.com', 'bob@example.com', 'charlie@example.com']}
df = pd.DataFrame(data)

# Define the table name
table_name = 'users'

# Method to upsert
def upsert_dataframe(df, table_name, engine):
    # Convert the DataFrame to a list of dictionaries for SQLAlchemy to use in the raw SQL statement
    records = df.to_dict(orient='records')

    # Define the insert statement with an ON CONFLICT clause for upsert functionality
    insert_stmt = f"""
    INSERT INTO {table_name} (user_id, username, email)
    VALUES (:user_id, :username, :email)
    ON CONFLICT (user_id) DO UPDATE SET 
        username = EXCLUDED.username,
        email = EXCLUDED.email;
    """
    
    # Execute the upsert
    with engine.connect() as conn:
        conn.execute(text(insert_stmt), records)

# Perform the upsert
upsert_dataframe(df, table_name, engine)
```

### PR DESCRIPTION
**Pull request by bh2smith**
Introducing `INSERT ... ON CONFLICT ... DO UPDATE`!

This works well but also requires primary key or, more loosely, a uniqueness constraint on the table in question. 
We have insightful error messaging that will literally print out the table alterations require for the provided `conflict_columns`:

So, when table exists without index we will see this error log suggesting how to fix the issue.
```sh
The ON CONFLICT clause requires a unique or exclusion constraint on the column(s): hash. Please ensure the table 'results_4238114' has the necessary constraint.To fix this, you can run the following SQL command:
ALTER TABLE results_4238114 ADD CONSTRAINT results_4238114_hash_unique UNIQUE (hash);
```

Note that, additionally, the table does not need to exist first. It will be created without any validation on the uniqueness constraints at first run. We only validate the uniqueness at the first time of actual upsert.

## Test Plan 

For now this is

1. Run it 
```
docker-compose up -d
make run
```
If you already have the table (from previous runs) you will hit the error above.
Execute the suggested ALTER TABLE command and try again!

**UPDATE:** 
There are now "unit" tests of all conflict column validation and upsert method works as expected.

Closes #56.

### CODE REVIEW

File: .github/workflows/coverage.yaml [hunk 1]
```diff
@@ -31,7 +31,7 @@ jobs:
       - name: Install Requirements
         run: POETRY_VIRTUALENVS_CREATE=false python -m poetry install
       - name: Run tests with coverage
-        run: pytest --cov=src --cov-report=xml --cov-fail-under=80
+        run: pytest --cov=src --cov-report=xml --cov-fail-under=90
```
Comment (mooster531) [inline .github/workflows/coverage.yaml]: :rocket:

File: Makefile [hunk 2]
```diff
@@ -18,6 +18,7 @@ fmt:
 
 lint:
 	ruff check .
+	pylint src/
```
Comment (mooster531) [inline Makefile:21]: Completely off-topic but employing _both_ pylint and ruff seems a little overkill. I'm not necessarily opposed to it, just seems weird to have two linters.

File: src/destinations/postgres.py [hunk 3]
```diff
@@ -3,12 +3,13 @@
 from typing import Literal
 
 import sqlalchemy
-from sqlalchemy import create_engine
+from sqlalchemy import create_engine, inspect
+from sqlalchemy.sql import text
```
Comment (mooster531) [inline src/destinations/postgres.py]:
```suggestion
from sqlalchemy import MetaData, Table, create_engine, inspect
from sqlalchemy.dialects.postgresql import insert
```

File: src/destinations/postgres.py [hunk 4]
```diff
@@ -75,17 +119,91 @@ def save(
...
+            {update_assignments};
+        """
+
+        # Convert the DataFrame to a list of dictionaries for SQLAlchemy to use
+        records = df.to_dict(orient="records")
+
+        # Execute the upsert
+        with self.engine.connect() as conn:
+            with conn.begin():
+                conn.execute(text(insert_stmt), records)
```
Comment (mooster531) [inline src/destinations/postgres.py]:
```suggestion
    def upsert(self, data: TypedDataFrame) -> None:
        """Upsert data from a DataFrame into a SQL table.

        :param data: Typed pandas DataFrame containing the data to upsert.
        """
        if not self.table_exists():
            # Do append.
            self.append(data)
            return

        self.validate_unique_constraints()
        df, _ = data
        # Get all column names from the DataFrame
        columns = df.columns.tolist()

        metadata = MetaData()
        table = Table(self.table_name, metadata, autoload_with=self.engine)
        statement = insert(table).values(**{col: df[col] for col in columns})

        statement = statement.on_conflict_do_update(
            index_elements=self.index_columns,
            set_={col: getattr(statement.excluded, col) for col in columns},
        )
        records = df.to_dict(orient="records")
        with self.engine.connect() as conn:
            with conn.begin():
                conn.execute(statement, records)

```

**Review by mooster531 — CHANGES_REQUESTED** [review]
I think that if we're using an ORM, we should use the ORM as much as we can :) I simplified the upsert function using sqlalchemy's built-in "on conflict update" utility.

File: Makefile (hunk 2 above)
Comment (bh2smith) [reply Makefile:21]: Pylint runs in our CI. It was failing cuz they were out of sync. I would rather catch them all till ruff catches up.

**Comment by bh2smith** [comment]
I think your comments have all been addressed.

**Review by mooster531 — APPROVED** [review]

**Comment by mooster531** [comment]
🚀🚀🚀

**Pull request merged**